File: my-folder/1316-fizz-buzz-multithreaded/solution.java
import java.util.concurrent.Semaphore;
class FizzBuzz {
    private int n;
    private boolean done;
    private Semaphore num_sema;
    private Semaphore fizz_sema;
    private Semaphore buzz_sema;
    private Semaphore fizzbuzz_sema;

    public FizzBuzz(int n) {
        this.n = n;
        num_sema = new Semaphore(1);
        fizz_sema = new Semaphore(0);
        buzz_sema = new Semaphore(0);
        fizzbuzz_sema = new Semaphore(0);
        this.done = false;
    }

    // printFizz.run() outputs "fizz".
    public void fizz(Runnable printFizz) throws InterruptedException {
        while (true){
            this.fizz_sema.acquire();
            if (this.done){
                break;
            }
            printFizz.run();
            this.num_sema.release();
        }
    }

    // printBuzz.run() outputs "buzz".
    public void buzz(Runnable printBuzz) throws InterruptedException {
        while (true){
            this.buzz_sema.acquire();
            if (this.done){
                break;
            }
            printBuzz.run();
            this.num_sema.release();
        }
    }

    // printFizzBuzz.run() outputs "fizzbuzz".
    public void fizzbuzz(Runnable printFizzBuzz) throws InterruptedException {
        while (true){
            this.fizzbuzz_sema.acquire();
            if (this.done){
                break;
            }
            printFizzBuzz.run();
            this.num_sema.release();
        }
    }

    // printNumber.accept(x) outputs "x", where x is an integer.
    public void number(IntConsumer printNumber) throws InterruptedException {
        
        for (int i=1; i<this.n+1; i++){
            this.num_sema.acquire();

            if (i % 15 == 0){
                this.fizzbuzz_sema.release();
            }
            else if (i % 3 == 0){
                this.fizz_sema.release();
            }
            else if (i % 5 == 0){
                this.buzz_sema.release();
            }
            else {
                printNumber.accept(i);
                this.num_sema.release();
            }
        }

        this.num_sema.acquire();
        this.done = true;
        this.num_sema.release();
        this.fizzbuzz_sema.release();
        this.fizz_sema.release();
        this.buzz_sema.release();
    }
}

File: my-folder/1316-fizz-buzz-multithreaded/solution.py
import threading

class FizzBuzz:
    def __init__(self, n: int):
        self.n = n
        self.i = 0
        self.lock = threading.Lock()
        self.num_lock = threading.Semaphore(1)
        self.fizz_lock = threading.Semaphore(0)
        self.buzz_lock = threading.Semaphore(0)
        self.fizzbuzz_lock = threading.Semaphore(0)

    def fizz(self, printFizz: 'Callable[[], None]') -> None:
        while True:
            self.fizz_lock.acquire()
            if self.i == self.n:
                break
            printFizz()
            self.num_lock.release()

    def buzz(self, printBuzz: 'Callable[[], None]') -> None:
        while True:
            self.buzz_lock.acquire()
            if self.i == self.n:
                break
            printBuzz()
            self.num_lock.release()

    def fizzbuzz(self, printFizzBuzz: 'Callable[[], None]') -> None:
        while True:
            self.fizzbuzz_lock.acquire()
            if self.i == self.n:
                break
            printFizzBuzz()
            self.num_lock.release()

    def number(self, printNumber: 'Callable[[int], None]') -> None:

        for i in range(1, self.n + 1):
            self.num_lock.acquire()
            
            if i % 15 == 0:
                self.fizzbuzz_lock.release()
            elif i % 3 == 0:
                self.fizz_lock.release()
            elif i % 5 == 0:
                self.buzz_lock.release()
            else:
                # Print a normal number => release num_lock here
                printNumber(i)
                self.num_lock.release()

        self.num_lock.acquire()
        self.i = self.n
        self.num_lock.release()
        self.fizz_lock.release()
        self.buzz_lock.release()
        self.fizzbuzz_lock.release()

